// polyserve/lib/start_server.js
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
/// <reference path="../custom_typings/spdy.d.ts" />
const assert = require("assert");
const express = require("express");
const fs = require("mz/fs");
const path = require("path");
const send = require("send");
// TODO: Switch to node-http2 when compatible with express
// https://github.com/molnarg/node-http2/issues/100
const http = require("spdy");
const bower_config_1 = require("./bower_config");
const make_app_1 = require("./make_app");
const next_open_port_1 = require("./util/next_open_port");
const open_browser_1 = require("./util/open_browser");
const push_1 = require("./util/push");
const tls_1 = require("./util/tls");
function applyDefaultOptions(options) {
    return __awaiter(this, void 0, void 0, function* () {
        const withDefaults = Object.assign({}, options);
        Object.assign(withDefaults, {
            port: yield next_open_port_1.nextOpenPort(options.port),
            hostname: options.hostname || 'localhost',
            root: path.resolve(options.root || '.'),
            certPath: options.certPath || 'cert.pem',
            keyPath: options.keyPath || 'key.pem',
        });
        withDefaults.packageName = options.packageName ||
            bower_config_1.bowerConfig(withDefaults.root).name || path.basename(process.cwd());
        return withDefaults;
    });
}
/**
 * @return {Promise} A Promise that completes when the server has started.
 * @deprecated Please use `startServers` instead. This function will be removed
 *     in a future release.
 */
function startServer(options) {
    return __awaiter(this, void 0, void 0, function* () {
        return (yield _startServer(options)).server;
    });
}
exports.startServer = startServer;
function _startServer(options) {
    return __awaiter(this, void 0, void 0, function* () {
        options = options || {};
        assertNodeVersion(options);
        try {
            const fullOptions = yield applyDefaultOptions(options);
            const app = getApp(options);
            const server = yield startWithPort(fullOptions, app);
            return { app, server };
        }
        catch (e) {
            console.error('ERROR: Server failed to start:', e);
            throw new Error(e);
        }
    });
}
/**
 * Starts one or more web servers, based on the given options and
 * variant bower_components directories that are found in the root dir.
 */
function startServers(options) {
    return __awaiter(this, void 0, void 0, function* () {
        const variants = yield findVariants(options);
        // TODO(rictic): support manually configuring variants? tracking more
        //   metadata about them besides their names?
        if (variants.length > 0) {
            return yield startVariants(options, variants);
        }
        const serverAndApp = yield _startServer(options);
        return {
            options,
            kind: 'mainline',
            server: serverAndApp.server,
            app: serverAndApp.app,
        };
    });
}
exports.startServers = startServers;
function findVariants(options) {
    return __awaiter(this, void 0, void 0, function* () {
        const root = options.root || process.cwd();
        const filesInRoot = yield fs.readdir(root);
        const variants = filesInRoot
            .map(f => {
            const match = f.match(/^bower_components-(.*)/);
            return match && { name: match[1], directory: match[0] };
        })
            .filter(f => f != null && f.name !== '');
        return variants;
    });
}
function startVariants(options, variants) {
    return __awaiter(this, void 0, void 0, function* () {
        const mainlineOptions = Object.assign({}, options);
        mainlineOptions.port = 0;
        const mainServer = yield _startServer(mainlineOptions);
        const mainServerInfo = {
            kind: 'mainline',
            server: mainServer.server,
            app: mainServer.app,
            options: mainlineOptions,
        };
        const variantServerInfos = [];
        for (const variant of variants) {
            const variantOpts = Object.assign({}, options);
            variantOpts.port = 0;
            variantOpts.componentDir = variant.directory;
            const variantServer = yield _startServer(variantOpts);
            variantServerInfos.push({
                kind: 'variant',
                variantName: variant.name,
                dependencyDir: variant.directory,
                server: variantServer.server,
                app: variantServer.app,
                options: variantOpts
            });
        }
        ;
        const controlServerInfo = yield startControlServer(options, mainServerInfo, variantServerInfos);
        const servers = [controlServerInfo, mainServerInfo]
            .concat(variantServerInfos);
        return {
            kind: 'MultipleServers',
            control: controlServerInfo,
            mainline: mainServerInfo,
            variants: variantServerInfos, servers,
        };
    });
}
function startControlServer(options, mainlineInfo, variantInfos) {
    return __awaiter(this, void 0, void 0, function* () {
        const fullOptions = yield applyDefaultOptions(options);
        const app = express();
        app.get('/api/serverInfo', (_req, res) => {
            res.contentType('json');
            res.send(JSON.stringify({
                packageName: fullOptions.packageName,
                mainlineServer: {
                    port: mainlineInfo.server.address().port,
                },
                variants: variantInfos.map(info => {
                    return { name: info.variantName, port: info.server.address().port };
                })
            }));
            res.end();
        });
        const indexPath = path.join(__dirname, '..', 'static', 'index.html');
        app.get('/', (_req, res) => __awaiter(this, void 0, void 0, function* () {
            res.contentType('html');
            const indexContents = yield fs.readFile(indexPath, 'utf-8');
            res.send(indexContents);
            res.end();
        }));
        const controlServer = {
            kind: 'control',
            options: fullOptions,
            server: yield startWithPort(fullOptions, app), app
        };
        return controlServer;
    });
}
const portInUseMessage = (port) => `
ERROR: Port in use: ${port}
Please choose another port, or let an unused port be chosen automatically.
`;
function getApp(options) {
    // Preload the h2-push manifest to avoid the cost on first push
    if (options.pushManifestPath) {
        push_1.getPushManifest(options.root, options.pushManifestPath);
    }
    const root = options.root || '.';
    const app = express();
    const polyserve = make_app_1.makeApp({
        componentDir: options.componentDir,
        packageName: options.packageName,
        root: root,
        compile: options.compile,
    });
    options.packageName = polyserve.packageName;
    const filePathRegex = /.*\/.+\..{1,}$/;
    app.use('/components/', polyserve);
    app.get('/*', (req, res) => {
        push_1.pushResources(options, req, res);
        const filePath = req.path;
        send(req, filePath, { root: root })
            .on('error', (error) => {
            if ((error).status === 404 && !filePathRegex.test(filePath)) {
                send(req, '/', { root: root }).pipe(res);
            }
            else {
                res.statusCode = error.status || 500;
                res.end(error.message);
            }
        })
            .pipe(res);
    });
    return app;
}
exports.getApp = getApp;
/**
 * Determines whether a protocol requires HTTPS
 * @param {string} protocol Protocol to evaluate.
 * @returns {boolean}
 */
function isHttps(protocol) {
    return ['https/1.1', 'https', 'h2'].indexOf(protocol) > -1;
}
/**
 * Gets the URLs for the main and component pages
 * @param {ServerOptions} options
 * @returns {{serverUrl: {protocol: string, hostname: string, port: string},
 * componentUrl: url.Url}}
 */
function getServerUrls(options, server) {
    const address = server.address();
    const serverUrl = {
        protocol: isHttps(options.protocol) ? 'https' : 'http',
        hostname: address.address,
        port: String(address.port),
    };
    const componentUrl = Object.assign({}, serverUrl);
    componentUrl.pathname = `components/${options.packageName}/`;
    return { serverUrl, componentUrl };
}
exports.getServerUrls = getServerUrls;
/**
 * Asserts that Node version is valid for h2 protocol
 * @param {ServerOptions} options
 */
function assertNodeVersion(options) {
    if (options.protocol === 'h2') {
        const matches = /(\d+)\./.exec(process.version);
        if (matches) {
            const major = Number(matches[1]);
            assert(major >= 5, 'h2 requires ALPN which is only supported in node.js >= 5.0');
        }
    }
}
/**
 * Creates an HTTP(S) server
 * @param app
 * @param {ServerOptions} options
 * @returns {Promise<http.Server>} Promise of server
 */
function createServer(app, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const opt = { spdy: { protocols: [options.protocol] } };
        if (isHttps(options.protocol)) {
            const keys = yield tls_1.getTLSCertificate(options.keyPath, options.certPath);
            opt.key = keys.key;
            opt.cert = keys.cert;
        }
        else {
            opt.spdy.plain = true;
            opt.spdy.ssl = false;
        }
        return http.createServer(opt, app);
    });
}
/**
 * Starts an HTTP(S) server on a specific port
 * @param {ServerOptions} userOptions
 * @returns {Promise<http.Server>} Promise of server
 */
function startWithPort(options, app) {
    return __awaiter(this, void 0, void 0, function* () {
        const server = yield createServer(app, options);
        yield new Promise((resolve, reject) => {
            server.listen(options.port, options.hostname, () => {
                resolve();
            });
            server.on('error', (err) => {
                if (err.code === 'EADDRINUSE') {
                    console.error(portInUseMessage(options.port));
                }
                reject(err);
            });
        });
        const urls = getServerUrls(options, server);
        open_browser_1.openBrowser(options, urls.serverUrl, urls.componentUrl);
        return server;
    });
}
exports.startWithPort = startWithPort;
